packages/babel-traverse/src/scope/index.ts

```diff
@@ -359,7 +359,16 @@ export default class Scope {
   static contextVariables = ["arguments", "undefined", "Infinity", "NaN"];
 
   get parent() {
-    const parent = this.path.findParent(p => p.isScope());
+    let parent,
+      path = this.path;
+    do {
+      // Skip method scope if coming from inside computed key
+      const isKey = path.key === "key";
+      path = path.parentPath;
+      if (isKey && path.isMethod()) path = path.parentPath;
+      if (path && path.isScope()) parent = path;
+    } while (path && !parent);
+
```

Comment:
Q: Can we call `Scope#parent` recursively? 
```js
const scopeParent = this._parent()
if (this path key is key && its parentPath is Method) {
  return scopeParent.parent;
}
```

Reply:
I agree that would be clearer code. But, if I've understood correctly what you're suggesting, I don't think it can work.

Problem is that the when testing if to skip a scope, we need to know not just where we are, but where we've come from (from inside the key or not?).

The computed key may not be the initial path. e.g.:

```js
const a = 'outside';
const o = {
  [
    (function f() { return a; })() + ''
  ]() {
    const a = 'inside';
  }
};
```

If we start at the scope of `f()`, at that point the parent is not a method, so the check is not triggered at this point. We proceed down to the parent scope (with `this._parent()`) and then discover it's a method. But by this point we don't know if we came from inside the key or not - we don't have the previous path - so we don't know whether to skip it or not. To fix this, the method check has to move into `_parent()`, at which point `_parent()` becomes as complicated as the original.

Here's a couple of other versions. Do you prefer either of them?

```js
get parent() {
  let path = this.path;
  // eslint-disable-next-line no-constant-condition
  while (true) {
    // Skip method scope if coming from inside computed key
    path =
      path.key === "key" && path.parentPath.isMethod()
        ? path.parentPath.parentPath
        : path.parentPath;
    if (!path) return undefined;
    if (path.isScope()) return path.scope;
  }
}
```

```js
get parent() {
  // Skip method scope if coming from inside computed key
  let isKey = this.path.key === "key";
  const parent = this.path.findParent(p => {
    const isScope = p.isScope() && !(isKey && p.isMethod());
    isKey = p.key === "key";
    return isScope;
  });
  return parent?.scope;
}
```

Reply:
Thanks for a detailed explainer! I slightly prefer the second one but no strong opinion.